In [1]:
import pandas as pd
import numpy as np
from sklearn import linear_model
from sklearn.model_selection import train_test_split
from sklearn.metrics import mean_squared_error
from sklearn.feature_selection import SelectKBest, f_regression

Necessary libraries are downloaded. Pandas and numpy are standard data science libraries. Sklearn is for feature selection and prediction.

In [2]:
df = pd.read_csv('DelayedFlights.csv')

DelayedFlights.csv dataset is uploaded into a pandas dataframe.

In [3]:
df.head()

Unnamed: 0  Year  Month  DayofMonth  DayOfWeek  DepTime  CRSDepTime  \
0           0  2008      1           3          4   2003.0        1955   
1           1  2008      1           3          4    754.0         735   
2           2  2008      1           3          4    628.0         620   
3           4  2008      1           3          4   1829.0        1755   
4           5  2008      1           3          4   1940.0        1915   

   ArrTime  CRSArrTime UniqueCarrier  ...  TaxiIn TaxiOut  Cancelled  \
0   2211.0        2225            WN  ...     4.0     8.0          0   
1   1002.0        1000            WN  ...     5.0    10.0          0   
2    804.0         750            WN  ...     3.0    17.0          0   
3   1959.0        1925            WN  ...     3.0    10.0          0   
4   2121.0        2110            WN  ...     4.0    10.0          0   

   CancellationCode  Diverted  CarrierDelay  WeatherDelay NASDelay  \
0                 N         0           NaN           NaN      NaN   
1                 N         0           NaN           NaN      NaN   
2                 N         0           NaN           NaN      NaN   
3                 N         0           2.0           0.0      0.0   
4                 N         0           NaN           NaN      NaN   

  SecurityDelay  LateAircraftDelay  
0           NaN                NaN  
1           NaN                NaN  
2           NaN                NaN  
3           0.0               32.0  
4           NaN                NaN  

[5 rows x 30 columns]

Exploratory Data Analysis. The first 5 rows of the table

In [4]:
df.describe()

Unnamed: 0       Year         Month    DayofMonth     DayOfWeek  \
count  1.936758e+06  1936758.0  1.936758e+06  1.936758e+06  1.936758e+06   
mean   3.341651e+06     2008.0  6.111106e+00  1.575347e+01  3.984827e+00   
std    2.066065e+06        0.0  3.482546e+00  8.776272e+00  1.995966e+00   
min    0.000000e+00     2008.0  1.000000e+00  1.000000e+00  1.000000e+00   
25%    1.517452e+06     2008.0  3.000000e+00  8.000000e+00  2.000000e+00   
50%    3.242558e+06     2008.0  6.000000e+00  1.600000e+01  4.000000e+00   
75%    4.972467e+06     2008.0  9.000000e+00  2.300000e+01  6.000000e+00   
max    7.009727e+06     2008.0  1.200000e+01  3.100000e+01  7.000000e+00   

            DepTime    CRSDepTime       ArrTime    CRSArrTime     FlightNum  \
count  1.936758e+06  1.936758e+06  1.929648e+06  1.936758e+06  1.936758e+06   
mean   1.518534e+03  1.467473e+03  1.610141e+03  1.634225e+03  2.184263e+03   
std    4.504853e+02  4.247668e+02  5.481781e+02  4.646347e+02  1.944702e+03   
min    1.000000e+00  0.000000e+00  1.000000e+00  0.000000e+00  1.000000e+00   
25%    1.203000e+03  1.135000e+03  1.316000e+03  1.325000e+03  6.100000e+02   
50%    1.545000e+03  1.510000e+03  1.715000e+03  1.705000e+03  1.543000e+03   
75%    1.900000e+03  1.815000e+03  2.030000e+03  2.014000e+03  3.422000e+03   
max    2.400000e+03  2.359000e+03  2.400000e+03  2.400000e+03  9.742000e+03   

       ...      Distance        TaxiIn       TaxiOut     Cancelled  \
count  ...  1.936758e+06  1.929648e+06  1.936303e+06  1.936758e+06   
mean   ...  7.656862e+02  6.812975e+00  1.823220e+01  3.268348e-04   
std    ...  5.744797e+02  5.273595e+00  1.433853e+01  1.807562e-02   
min    ...  1.100000e+01  0.000000e+00  0.000000e+00  0.000000e+00   
25%    ...  3.380000e+02  4.000000e+00  1.000000e+01  0.000000e+00   
50%    ...  6.060000e+02  6.000000e+00  1.400000e+01  0.000000e+00   
75%    ...  9.980000e+02  8.000000e+00  2.100000e+01  0.000000e+00   
max    ...  4.962000e+03  2.400000e+02  4.220000e+02  1.000000e+00   

           Diverted  CarrierDelay  WeatherDelay      NASDelay  SecurityDelay  \
count  1.936758e+06  1.247488e+06  1.247488e+06  1.247488e+06   1.247488e+06   
mean   4.003598e-03  1.917940e+01  3.703571e+00  1.502164e+01   9.013714e-02   
std    6.314722e-02  4.354621e+01  2.149290e+01  3.383305e+01   2.022714e+00   
min    0.000000e+00  0.000000e+00  0.000000e+00  0.000000e+00   0.000000e+00   
25%    0.000000e+00  0.000000e+00  0.000000e+00  0.000000e+00   0.000000e+00   
50%    0.000000e+00  2.000000e+00  0.000000e+00  2.000000e+00   0.000000e+00   
75%    0.000000e+00  2.100000e+01  0.000000e+00  1.500000e+01   0.000000e+00   
max    1.000000e+00  2.436000e+03  1.352000e+03  1.357000e+03   3.920000e+02   

       LateAircraftDelay  
count       1.247488e+06  
mean        2.529647e+01  
std         4.205486e+01  
min         0.000000e+00  
25%         0.000000e+00  
50%         8.000000e+00  
75%         3.300000e+01  
max         1.316000e+03  

[8 rows x 25 columns]

Exploratory Data Analysis. Count, mean, median values of each column. 

In [5]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1936758 entries, 0 to 1936757
Data columns (total 30 columns):
Unnamed: 0           int64
Year                 int64
Month                int64
DayofMonth           int64
DayOfWeek            int64
DepTime              float64
CRSDepTime           int64
ArrTime              float64
CRSArrTime           int64
UniqueCarrier        object
FlightNum            int64
TailNum              object
ActualElapsedTime    float64
CRSElapsedTime       float64
AirTime              float64
ArrDelay             float64
DepDelay             float64
Origin               object
Dest                 object
Distance             int64
TaxiIn               float64
TaxiOut              float64
Cancelled            int64
CancellationCode     object
Diverted             int64
CarrierDelay         float64
WeatherDelay         float64
NASDelay             float64
SecurityDelay        float64
LateAircraftDelay    float64
dtypes: float64(14), int64(11), object(5)
me

Exploratory Data Analysis. Data type of each column.

In [6]:
df = df.select_dtypes(exclude=['object'])

Object-type columns are excluded because they are irrelevant and it is difficult to use feature selection algorithms with them.

In [7]:
df.drop(['Unnamed: 0'], axis=1, inplace=True)

The first column is dropped since it is not well defined, column name is not set.

In [8]:
df.shape

(1936758, 24)

The number of rows and columns are checked to see if cleansing is successfull. 6 columns are dropped. Irrelevant columns are dropped successfully. 

In [9]:
df.isnull().sum()

Year                      0
Month                     0
DayofMonth                0
DayOfWeek                 0
DepTime                   0
CRSDepTime                0
ArrTime                7110
CRSArrTime                0
FlightNum                 0
ActualElapsedTime      8387
CRSElapsedTime          198
AirTime                8387
ArrDelay               8387
DepDelay                  0
Distance                  0
TaxiIn                 7110
TaxiOut                 455
Cancelled                 0
Diverted                  0
CarrierDelay         689270
WeatherDelay         689270
NASDelay             689270
SecurityDelay        689270
LateAircraftDelay    689270
dtype: int64

Null values are detected.

In [10]:
df.dropna(axis=0, how='any', inplace=True)

Rows which include null values are dropped. 

In [11]:
df.isnull().sum()

Year                 0
Month                0
DayofMonth           0
DayOfWeek            0
DepTime              0
CRSDepTime           0
ArrTime              0
CRSArrTime           0
FlightNum            0
ActualElapsedTime    0
CRSElapsedTime       0
AirTime              0
ArrDelay             0
DepDelay             0
Distance             0
TaxiIn               0
TaxiOut              0
Cancelled            0
Diverted             0
CarrierDelay         0
WeatherDelay         0
NASDelay             0
SecurityDelay        0
LateAircraftDelay    0
dtype: int64

Null values are checked again to see if dropna works successfully. So far so good.

In [12]:
df.reset_index(drop=True, inplace=True)

We need to rearrange the index after some rows are dropped.

In [13]:
df.head()

Year  Month  DayofMonth  DayOfWeek  DepTime  CRSDepTime  ArrTime  \
0  2008      1           3          4   1829.0        1755   1959.0   
1  2008      1           3          4   1937.0        1830   2037.0   
2  2008      1           3          4   1644.0        1510   1845.0   
3  2008      1           3          4   1452.0        1425   1640.0   
4  2008      1           3          4   1323.0        1255   1526.0   

   CRSArrTime  FlightNum  ActualElapsedTime  ...  Distance  TaxiIn  TaxiOut  \
0        1925       3920               90.0  ...       515     3.0     10.0   
1        1940        509              240.0  ...      1591     3.0      7.0   
2        1725       1333              121.0  ...       828     6.0      8.0   
3        1625        675              228.0  ...      1489     7.0      8.0   
4        1510          4              123.0  ...       838     4.0      9.0   

   Cancelled  Diverted  CarrierDelay  WeatherDelay  NASDelay  SecurityDelay  \
0          0         0           2.0           0.0       0.0            0.0   
1          0         0          10.0           0.0       0.0            0.0   
2          0         0           8.0           0.0       0.0            0.0   
3          0         0           3.0           0.0       0.0            0.0   
4          0         0           0.0           0.0       0.0            0.0   

   LateAircraftDelay  
0               32.0  
1               47.0  
2               72.0  
3               12.0  
4               16.0  

[5 rows x 24 columns]

First 5 rows are checked to see if reindexing is successful.

In [14]:
y=df.loc[:,'DepDelay']
X=df.loc[:, df.columns != 'DepDelay']

Target variable is assigned. 'DepDelay' column is the target varible and rest of the columns will be used in a feature selection algorithm to select most relevant features.

In [15]:
np.seterr(divide='ignore', invalid='ignore')
X_selected_features=SelectKBest(f_regression, k=10)
X_new=X_selected_features.fit_transform(X,y)


Both input and output is numerical in the problem that we try to solve, so pearson's correlation method seems to best fit for our model. f_regression algorithm implements this method.

In [16]:
cols = X_selected_features.get_support()

The result of feature selection is assigned into a variable called cols.

In [17]:
df_ready = X.iloc[:,cols]

Cleansing and feature selection are done, our dataset is ready to train.

In [18]:
df_ready.shape

(1247488, 10)

Exploring the last version of our dataset. 10 columns and 1247488 rows.

In [19]:
df_ready.head()

DepTime  CRSDepTime  ArrTime  CRSArrTime  FlightNum  ArrDelay  \
0   1829.0        1755   1959.0        1925       3920      34.0   
1   1937.0        1830   2037.0        1940        509      57.0   
2   1644.0        1510   1845.0        1725       1333      80.0   
3   1452.0        1425   1640.0        1625        675      15.0   
4   1323.0        1255   1526.0        1510          4      16.0   

   CarrierDelay  WeatherDelay  NASDelay  LateAircraftDelay  
0           2.0           0.0       0.0               32.0  
1          10.0           0.0       0.0               47.0  
2           8.0           0.0       0.0               72.0  
3           3.0           0.0       0.0               12.0  
4           0.0           0.0       0.0               16.0

First 5 rows are seen after feature selection and cleansing.

In [20]:
df_ready.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1247488 entries, 0 to 1247487
Data columns (total 10 columns):
DepTime              1247488 non-null float64
CRSDepTime           1247488 non-null int64
ArrTime              1247488 non-null float64
CRSArrTime           1247488 non-null int64
FlightNum            1247488 non-null int64
ArrDelay             1247488 non-null float64
CarrierDelay         1247488 non-null float64
WeatherDelay         1247488 non-null float64
NASDelay             1247488 non-null float64
LateAircraftDelay    1247488 non-null float64
dtypes: float64(7), int64(3)
memory usage: 95.2 MB


Data type of each column after preprocessing.

In [21]:
X_train, X_test, y_train, y_test = train_test_split(X_new, y, test_size=0.2, random_state=42)

Test data and train data are splitted. random_state is used to get same result if code is implemented later again. 

In [22]:
reg= linear_model.LinearRegression()

Regressor is created.

In [23]:
reg.fit(X_train, y_train)

LinearRegression(copy_X=True, fit_intercept=True, n_jobs=None, normalize=False)

Training is implemented.

In [24]:
y_pred = reg.predict(X_test)

Prediction is implemented on test data.

In [25]:
rmse = np.sqrt(mean_squared_error(y_test, y_pred))

calculation of average value error

In [26]:
rmse

16.31499947765962

average value error

In [27]:
reg.score(X_test, y_test)

0.9268409765493003

calculation of r2
